test/python/test_spectral_filter.py

```diff
@@ -0,0 +1,331 @@
+        """ set the list of amplitudes of the sin waves """
+        self.amplitudes = amps
+
+    def set_frequencies(self, freqs):
+        """ set the list of frequencies of the sin waves"""
+        self.frequencies = freqs
+
+    def print_status(self, msg):
+        if self.verbose:
+            rank = MPI.COMM_WORLD.Get_rank()
+            sys.stderr.write('[%d] generate_data::%s\n'%(rank, msg))
+
+    def get_point_array_names(self):
+        return ['f_t']
+
+    def report(self, port, md_in):
+        self.print_status('report(override)')
+
+        # report arrays we generate
+        md_out = teca_metadata(md_in[0])
+        try:
+            arrays = md_out['arrays']
+        except:
+            arrays = []
+        md_out['arrays'] = arrays + ['f_t']
+
+        # get the extent of the dataset
+        wext = md_out['whole_extent']
+        ncells = (wext[1] - wext[0] + 1)* \
+                 (wext[3] - wext[2] + 1)*(wext[5] - wext[4] + 1)
+
+        # create the metadata for the writer
+        faa = teca_array_attributes(
+            teca_double_array_code.get(),
+            teca_array_attributes.point_centering,
+            int(ncells), 'none', 'f(t)',
+            'function of time',
+            None)
+
+        fatts = faa.to_metadata()
+        fatts['have_mesh_dim'] = [1,1,1,1]
+        fatts['mesh_dim_active'] = [1,1,1,1]
+
+        # put it in the array attributes
+        try:
+            atts = md_out['attributes']
+        except:
+            atts = teca_metadata()
+
+        atts['f_t'] = fatts
+
+        md_out['attributes'] = atts
+        return md_out
+
+    def execute(self, port, data_in, req_in):
+
+        # get the device to execute on
+        dev = -1
+        npmod = numpy
+        if have_cuda:
+            dev = req_in['device_id']
+            if dev >= 0:
+                npmod = cupy
+                cupy.cuda.Device(dev).use()
+
+        # get the input
+        mesh_in = as_const_teca_cartesian_mesh(data_in[0])
+
+        # get the time units in seconds
+        t_units = mesh_in.get_time_units()
+        seconds_per = get_number_of_seconds(t_units)
+
+        # get mesh dims and coordinate arrays
+        ext = mesh_in.get_extent()
+        nx = ext[3] - ext[2] + 1
+        ny = ext[1] - ext[0] + 1
+        nz = ext[5] - ext[4] + 1
+
+        # get the time step size
+        t_ext = mesh_in.get_temporal_extent()
+        nt = t_ext[1] - t_ext[0] + 1
+
+        t_bds = mesh_in.get_temporal_bounds()
+        dt = seconds_per * ( t_bds[1] - t_bds[0] ) / ( nt - 1 )
+        t0 = t_bds[0] * seconds_per
+
+        # generate the output
+        f_t = npmod.zeros((nt,nz,ny,nx), dtype=npmod.float64) + self.bias
+
+        i = 0
+        while i < nt:
+            t = t0 + i*dt
+            nf = len(self.frequencies)
+            j = 0
+            while j < nf:
+                f = self.frequencies[j]
+                A = self.amplitudes[j]
+                f_t[i,:,:,:] += A * npmod.sin( 2*pi*f*t )
```

Comment:
It might be a good idea to make a test based on the example given in the [scipy.signal.butter documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.butter.html), as that is a very simple test case with an easy-to-understand result (only the 20 hz wave should survive).

Reply:
The regression test does exactly this although the defaults test in the parameter space we are interested in. For instance here is how to reproduce the numpy example:
```
$ python3 "/home/bloring/work/teca/teca/test/python/test_spectral_filter.py" "90" "45" "2" "1000" "1" "1" "1" "0" "0.1" "0.05" "0.6667" "seconds" "high_pass" "analytic" "/home/bloring/work/teca/TECA_data/test_spectral_filter_low_pass" "1"

$ display test_spectral_filter_high_pass.png
```

And here is the plot the test generated:
![test_spectral_filter_high_pass](https://user-images.githubusercontent.com/115053/181654727-3e6ace08-4cf6-434f-b9e2-11e386d6b07e.png)

Reply:
note: you can see from the filter's frequency response in this case  (image in a comment below) that 10Hz signal is not going to be completely filtered. I think this explains the error in the 20 Hz wave amplitude here.

Reply:
Cool!  I think this raises an important question though: why does the solution at the bottom of the [scipy.signal.butterworth](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.butter.html) documentation have no evidence of the 10 Hz signal?  It could be because your example uses a 2nd order filter (if I read the code correctly) and theirs uses a 10th-order filter.

Reply:
Yes, You're right, I used order 2. When I use order 10 I get a better result.
![test_spectral_filter_high_pass](https://user-images.githubusercontent.com/115053/181775509-c76bbb1a-6ddc-439b-918e-19178367c956.png)